lib/Greenspun/lib/Dicts.cpp

```diff
@@ -0,0 +1,365 @@
+  VPackObjectBuilder ob(&result);
+  for (auto&& pair : VPackArrayIterator(params)) {
+    if (pair.isArray() && pair.length() == 2) {
+      if (pair.at(0).isString()) {
+        result.add(pair.at(0).stringRef(), pair.at(1));
+        continue;
+      }
+    }
+
+    return EvalError("expected pairs of string and slice");
+  }
+  return {};
+}
+
+EvalResult Prim_DictKeys(Machine& ctx, VPackSlice const params, VPackBuilder& result) {
+  if (params.length() != 1) {
+    return EvalError("expected exactly one parameter");
+  }
+
+  auto obj = params.at(0);
+  if (!obj.isObject()) {
+    return EvalError("expected object, found: " + obj.toJson());
+  }
+
+  result.openArray();
+  for (VPackObjectIterator iter(params.at(0)); iter.valid(); iter++) {
+    result.add(iter.key());
+  }
+  result.close();
+
+  return {};
+}
+
+EvalResult Prim_DictDirectory(Machine& ctx, VPackSlice const params, VPackBuilder& result) {
+  if (params.length() != 1) {
+    return EvalError("expected exactly one parameter");
+  }
+
+  auto obj = params.at(0);
+  if (!obj.isObject()) {
+    return EvalError("expected object, found: " + obj.toJson());
+  }
+
+  std::list<std::list<std::string>> finalPaths;
+  std::list<std::string> currentPath;
+  createPaths(finalPaths, obj, currentPath);
+  pathToBuilder(finalPaths, result);
+
+  return {};
+}
+
+EvalResult MergeObjectSlice(VPackBuilder& result, VPackSlice const& sliceA,
+                            VPackSlice const& sliceB) {
+  VPackCollection::merge(result, sliceA, sliceB, true, false);
+  return {};
+}
+
+EvalResult Prim_MergeDict(Machine& ctx, VPackSlice const params, VPackBuilder& result) {
+  if (!params.isArray() && params.length() != 2) {
+    return EvalError("expected exactly two parameters");
+  }
+
+  if (!params.at(0).isObject()) {
+    return EvalError("expected object, found: " + params.at(0).toJson());
+  }
+  if (!params.at(1).isObject()) {
+    return EvalError("expected object, found: " + params.at(1).toJson());
+  }
+
+  MergeObjectSlice(result, params.at(0), params.at(1));
+  return {};
+}
+
+namespace {
+EvalResultT<VPackSlice> ReadAttribute(VPackSlice slice, VPackSlice key) {
+  if (!slice.isObject()) {
+    return EvalError("expect first parameter to be an object");
+  }
+
+  if (key.isString()) {
+    return slice.get(key.stringRef());
+  } else if (key.isArray()) {
+    struct Iter : VPackArrayIterator {
+      using VPackArrayIterator ::difference_type;
+      using value_type = VPackStringRef;
+      using VPackArrayIterator::iterator_category;
+      using VPackArrayIterator ::pointer;
+      using VPackArrayIterator ::reference;
+
+      value_type operator*() const {
+        return VPackArrayIterator::operator*().stringRef();
+      }
+      Iter begin() const { return Iter{VPackArrayIterator::begin()}; }
+      Iter end() const { return Iter{VPackArrayIterator::end()}; }
+    };
+
+    Iter i{VPackArrayIterator(key)};
+    return slice.get(i);
```

Comment:
I think it is correct here and will not fail (in case it is not expected to fail). Take a look at:
```
["attrib-ref", dict, key]
["attrib-ref", dict, path]
["attrib-ref-or", dict, key, default]
["attrib-ref-or", dict, path, default]
["attrib-ref-or-fail", dict, key]
["attrib-ref-or-fail", dict, path]
```

`*-or-fail` will do additional checking if key is not present. The "NoneSlice" will be not returned otherwise. Anyway, I've created a ticket to doublecheck.

Reply:
That's not the point. They all fail, and that's ok:
```
air> ["attrib-ref-or", {"bar": "baz"}, ["list"], 42]
error: uncaught exception with message: Invalid attribute path
at top-level
```

But they fail with an _uncaught exception_ instead of an `EvalError`, which should be avoided.